File: Plugin/src/main/java/com/bearmod/loader/permissions/PermissionManager.java
package com.bearmod.loader.permissions;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import android.util.Log;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * PHASE 7 STREAMLINED: Automated Permission Manager for BearMod integration
 * Focuses on programmatic confirmation with minimal user friction
 */
public class PermissionManager {
    private static final String TAG = "PermissionManager";
    
    // Required permissions for BearMod injection and PUBG interaction
    public static final String[] REQUIRED_PERMISSIONS = {
        Manifest.permission.WRITE_EXTERNAL_STORAGE,
        Manifest.permission.READ_EXTERNAL_STORAGE,
        Manifest.permission.SYSTEM_ALERT_WINDOW,
        Manifest.permission.ACCESS_NETWORK_STATE,
        Manifest.permission.INTERNET
    };

    // CRITICAL FIX: Android 8.0+ (API 26+) Unknown Sources permission
    public static final String REQUEST_INSTALL_PACKAGES = "android.permission.REQUEST_INSTALL_PACKAGES";
    
    // Permission display names (user-friendly)
    private static final Map<String, String> PERMISSION_DISPLAY_NAMES = new HashMap<>();
    private static final Map<String, String> PERMISSION_DESCRIPTIONS = new HashMap<>();
    
    static {
        // User-friendly permission names
        PERMISSION_DISPLAY_NAMES.put(Manifest.permission.WRITE_EXTERNAL_STORAGE, "Storage Access");
        PERMISSION_DISPLAY_NAMES.put(Manifest.permission.READ_EXTERNAL_STORAGE, "Storage Read");
        PERMISSION_DISPLAY_NAMES.put(Manifest.permission.SYSTEM_ALERT_WINDOW, "Overlay Permission");
        PERMISSION_DISPLAY_NAMES.put(Manifest.permission.ACCESS_NETWORK_STATE, "Network Access");
        PERMISSION_DISPLAY_NAMES.put(Manifest.permission.INTERNET, "Internet Access");
        
        // Concise permission purposes
        PERMISSION_DESCRIPTIONS.put(Manifest.permission.WRITE_EXTERNAL_STORAGE, "Store enhancement files");
        PERMISSION_DESCRIPTIONS.put(Manifest.permission.READ_EXTERNAL_STORAGE, "Access enhancement files");
        PERMISSION_DESCRIPTIONS.put(Manifest.permission.SYSTEM_ALERT_WINDOW, "Enable game overlay");
        PERMISSION_DESCRIPTIONS.put(Manifest.permission.ACCESS_NETWORK_STATE, "Check connectivity");
        PERMISSION_DESCRIPTIONS.put(Manifest.permission.INTERNET, "Download enhancements");
    }
    
    private final Context context;
    private PermissionStateListener permissionListener;
    private boolean isWaitingForPermissions = false;
    private String pendingWorkflow = null;
    
    public PermissionManager(Context context) {
        this.context = context.getApplicationContext();
    }
    
    /**
     * Check if all required permissions are granted
     */
    public boolean areAllPermissionsGranted() {
        for (String permission : REQUIRED_PERMISSIONS) {
            if (!isPermissionGranted(permission)) {
                return false;
            }
        }
        return true;
    }
    
    /**
     * Check if specific permission is granted
     */
    public boolean isPermissionGranted(String permission) {
        if (permission.equals(Manifest.permission.SYSTEM_ALERT_WINDOW)) {
            // Special handling for overlay permission
            return Settings.canDrawOverlays(context);
        } else if (permission.equals(REQUEST_INSTALL_PACKAGES)) {
            // CRITICAL FIX: Android 8.0+ (API 26+) Unknown Sources permission
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                return context.getPackageManager().canRequestPackageInstalls();
            }
            return true; // Not applicable for older Android versions
        } else {
            return ContextCompat.checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED;
        }
    }
    
    /**
     * Get list of missing permissions
     */
    public List<String> getMissingPermissions() {
        List<String> missing = new ArrayList<>();

        // Check standard permissions
        for (String permission : REQUIRED_PERMISSIONS) {
            if (!isPermissionGranted(permission)) {
                missing.add(permission);
            }
        }

        // CRITICAL FIX: Check Android 8.0+ REQUEST_INSTALL_PACKAGES permission
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (!isPermissionGranted(REQUEST_INSTALL_PACKAGES)) {
                missing.add(REQUEST_INSTALL_PACKAGES);
            }
        }

        return missing;
    }
    
    /**
     * Get permission status for UI display
     */
    public PermissionStatus getPermissionStatus(String permission) {
        if (isPermissionGranted(permission)) {
            return PermissionStatus.GRANTED;
        } else {
            return PermissionStatus.REQUIRED;
        }
    }
    
    /**
     * Get user-friendly permission name
     */
    public String getPermissionDisplayName(String permission) {
        return PERMISSION_DISPLAY_NAMES.getOrDefault(permission, permission);
    }
    
    /**
     * Get user-friendly permission description
     */
    public String getPermissionDescription(String permission) {
        return PERMISSION_DESCRIPTIONS.getOrDefault(permission, "Required for app functionality");
    }
    
    /**
     * STREAMLINED: Start automated permission request flow
     */
    public void requestPermissionsAutomated(Activity activity, String pendingWorkflow) {
        this.pendingWorkflow = pendingWorkflow;
        this.isWaitingForPermissions = true;

        List<String> missingPermissions = getMissingPermissions();
        if (missingPermissions.isEmpty()) {
            Log.d(TAG, "All permissions already granted");
            onPermissionsGranted();
            return;
        }

        Log.d(TAG, "Starting automated permission flow for " + missingPermissions.size() + " permissions");

        // Direct settings navigation without complex dialogs
        openPermissionSettings(activity, missingPermissions);
    }

    /**
     * LEGACY: Start manual permission request flow (maintained for compatibility)
     */
    public void startManualPermissionFlow(Activity activity, String pendingWorkflow) {
        requestPermissionsAutomated(activity, pendingWorkflow);
    }
    
    /**
     * STREAMLINED: Direct permission settings navigation
     */
    private void openPermissionSettings(Activity activity, List<String> missingPermissions) {
        try {
            // CRITICAL FIX: Prioritize REQUEST_INSTALL_PACKAGES permission for Android 8.0+
            boolean needsInstallPackagesPermission = missingPermissions.contains(REQUEST_INSTALL_PACKAGES);
            boolean needsOverlayPermission = missingPermissions.contains(Manifest.permission.SYSTEM_ALERT_WINDOW);

            Intent intent;
            String message;

            if (needsInstallPackagesPermission && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                // CRITICAL FIX: Direct to Unknown Sources permission for Android 8.0+
                intent = new Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES);
                intent.setData(Uri.parse("package:" + context.getPackageName()));
                message = "Enable 'Install unknown apps' permission";
            } else if (needsOverlayPermission) {
                // Direct to overlay permission
                intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION);
                intent.setData(Uri.parse("package:" + context.getPackageName()));
                message = "Enable overlay permission";
            } else {
                // Direct to app permissions
                intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                intent.setData(Uri.parse("package:" + context.getPackageName()));
                message = "Grant required permissions";
            }

            activity.startActivity(intent);

            if (permissionListener != null) {
                permissionListener.onPermissionFlowStarted(message);
            }

        } catch (Exception e) {
            Log.e(TAG, "Error opening settings", e);
            if (permissionListener != null) {
                permissionListener.onPermissionFlowFailed("Settings unavailable");
            }
        }
    }
    
    /**
     * Check permissions when user returns from settings
     */
    public void checkPermissionsOnResume() {
        if (!isWaitingForPermissions) {
            return;
        }
        
        Log.d(TAG, "Checking permissions on app resume...");
        
        if (areAllPermissionsGranted()) {
            Log.i(TAG, "All permissions granted!");
            isWaitingForPermissions = false;
            onPermissionsGranted();
        } else {
            List<String> stillMissing = getMissingPermissions();
            Log.d(TAG, "Still missing " + stillMissing.size() + " permissions");
            
            if (permissionListener != null) {
                permissionListener.onPermissionStatusChanged(stillMissing);
            }
        }
    }
    
    /**
     * Handle permissions granted
     */
    private void onPermissionsGranted() {
        if (permissionListener != null) {
            permissionListener.onAllPermissionsGranted();
            
            // Resume pending workflow if any
            if (pendingWorkflow != null) {
                permissionListener.onWorkflowResumed(pendingWorkflow);
                pendingWorkflow = null;
            }
        }
    }
    
    /**
     * Cancel permission flow
     */
    public void cancelPermissionFlow() {
        isWaitingForPermissions = false;
        pendingWorkflow = null;
        
        if (permissionListener != null) {
            permissionListener.onPermissionFlowCancelled();
        }
    }
    
    /**
     * Set permission state listener
     */
    public void setPermissionStateListener(PermissionStateListener listener) {
        this.permissionListener = listener;
    }
    
    /**
     * Check if waiting for permissions
     */
    public boolean isWaitingForPermissions() {
        return isWaitingForPermissions;
    }
    
    /**
     * Get pending workflow
     */
    public String getPendingWorkflow() {
        return pendingWorkflow;
    }

    /**
     * STREAMLINED: Programmatic permission confirmation
     * Returns true if all permissions are granted, false otherwise
     */
    public boolean confirmPermissionsQuickly() {
        List<String> missingPermissions = getMissingPermissions();
        boolean allGranted = missingPermissions.isEmpty();

        Log.d(TAG, "Quick permission check: " + (allGranted ? "GRANTED" : "MISSING " + missingPermissions.size()));

        return allGranted;
    }

    /**
     * STREAMLINED: Get brief permission status summary
     */
    public String getPermissionStatusSummary() {
        List<String> missingPermissions = getMissingPermissions();
        if (missingPermissions.isEmpty()) {
            return "All permissions granted";
        } else {
            // CRITICAL FIX: Provide specific guidance for REQUEST_INSTALL_PACKAGES
            if (missingPermissions.contains(REQUEST_INSTALL_PACKAGES)) {
                return "Install unknown apps permission required (Android 8.0+)";
            } else if (missingPermissions.contains(Manifest.permission.SYSTEM_ALERT_WINDOW)) {
                return "Overlay permission required";
            } else {
                return missingPermissions.size() + " permissions required";
            }
        }
    }
    
    /**
     * Permission status enumeration
     */
    public enum PermissionStatus {
        GRANTED,    // Permission is granted
        REQUIRED,   // Permission is required but not granted
        DENIED      // Permission was denied by user
    }
    
    /**
     * Permission state listener interface
     */
    public interface PermissionStateListener {
        void onPermissionFlowStarted(String message);
        void onAllPermissionsGranted();
        void onPermissionStatusChanged(List<String> missingPermissions);
        void onWorkflowResumed(String workflow);
        void onPermissionFlowFailed(String error);
        void onPermissionFlowCancelled();
    }
}

File: app/src/main/java/com/bearmod/PermissionManager.kt
package com.bearmod

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.Settings
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.core.net.toUri


/**
 * Comprehensive permission management utility for BearMod
 * 
 * Handles all permission requirements for package installation and OBB file management
 * across different Android versions with proper fallback mechanisms and user guidance.
 * 
 * Features:
 * - Unknown Sources permission for APK installation
 * - Storage permissions for OBB file access (API-level aware)
 * - Overlay permission integration
 * - Batch permission requests
 * - User-friendly permission explanations
 * - Thread-safe operations
 * 
 * @author BearMod Team
 * @version 1.0
 */
class PermissionManager private constructor(private val context: Context) {
    
    companion object {
        private const val TAG = "PermissionManager"
        
        // Request codes for permission handling
        const val REQUEST_UNKNOWN_SOURCES = 3001
        const val REQUEST_STORAGE_PERMISSION = 3002
        const val REQUEST_MANAGE_EXTERNAL_STORAGE = 3003
        const val REQUEST_OVERLAY_PERMISSION = 3004
        const val REQUEST_BATCH_PERMISSIONS = 3005
        
        // Permission groups for batch requests
        const val PERMISSION_GROUP_BASIC = "basic"
        const val PERMISSION_GROUP_STORAGE = "storage"
        const val PERMISSION_GROUP_INSTALLATION = "installation"
        const val PERMISSION_GROUP_ALL = "all"
        
        @SuppressLint("StaticFieldLeak")
        @Volatile
        private var INSTANCE: PermissionManager? = null
        
        /**
         * Get singleton instance of PermissionManager
         */
        fun getInstance(context: Context): PermissionManager {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: PermissionManager(context.applicationContext).also { INSTANCE = it }
            }
        }
    }
    
    /**
     * Data class to hold permission status information
     */
    data class PermissionStatus(
        val isGranted: Boolean,
        val permissionName: String,
        val rationale: String,
        val canShowRationale: Boolean = false,
        val requiresSystemSettings: Boolean = false
    )
    
    /**
     * Data class for batch permission results
     */
    data class BatchPermissionResult(
        val allGranted: Boolean,
        val grantedPermissions: List<String>,
        val deniedPermissions: List<String>,
        val permanentlyDeniedPermissions: List<String>
    )
    
    /**
     * Interface for permission result callbacks
     */
    interface PermissionCallback {
        fun onPermissionGranted(permission: String)
        fun onPermissionDenied(permission: String, canShowRationale: Boolean)
        fun onPermissionPermanentlyDenied(permission: String)
    }
    
    /**
     * Interface for batch permission callbacks
     */
    interface BatchPermissionCallback {
        fun onBatchPermissionResult(result: BatchPermissionResult)
    }
    
    // ================================================================================================
    // UNKNOWN SOURCES PERMISSION MANAGEMENT
    // ================================================================================================
    
    /**
     * Check if Unknown Sources permission is enabled for package installation
     * 
     * @return PermissionStatus with detailed information
     */
    fun checkUnknownSourcesPermission(): PermissionStatus {
        val isGranted =
            context.packageManager.canRequestPackageInstalls()

        return PermissionStatus(
            isGranted = isGranted,
            permissionName = "INSTALL_UNKNOWN_APPS",
            rationale = "Required to install APK files for game packages",
            requiresSystemSettings = true
        )
    }
    
    /**
     * Create intent to request Unknown Sources permission
     * 
     * @return Intent for requesting permission
     */
    fun createUnknownSourcesPermissionIntent(context: Context): Intent {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            // For Android 8.0 (Oreo) and above (API 26+)
            Log.d("PermissionManager", "Creating intent for ACTION_MANAGE_UNKNOWN_APP_SOURCES")
            Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES, Uri.parse("package:${context.packageName}"))
        } else {
            // For Android versions prior to 8.0 (Oreo)
            // Unknown Sources is a global setting in older versions, not app-specific.
            // It's typically found under Security settings.
            Log.d("PermissionManager", "Creating intent for ACTION_SECURITY_SETTINGS (pre-Oreo)")
            Intent(Settings.ACTION_SECURITY_SETTINGS)
        }
    }
    
    /**
     * Request Unknown Sources permission with proper handling
     * 
     * @param activity Activity context for launching permission request
     * @param requestCode Request code for result handling
     */
    fun requestUnknownSourcesPermission(activity: Activity, requestCode: Int = REQUEST_UNKNOWN_SOURCES) {
        try {
            // 1. Call the function with 'activity' as the Context.
            // 2. Remove ': Unit' from the declaration of 'intent'.
            val intent = createUnknownSourcesPermissionIntent(activity) // Pass activity as context

            activity.startActivityForResult(intent, requestCode)
            Log.d(TAG, "Unknown Sources permission request launched")
        } catch (e: Exception) {
            Log.e(TAG, "Error requesting Unknown Sources permission", e)
        }
    }
    
    // ================================================================================================
    // STORAGE PERMISSION MANAGEMENT
    // ================================================================================================
    
    /**
     * Check storage permissions based on Android version
     * 
     * @return PermissionStatus with version-appropriate storage permission info
     */
    fun checkStoragePermission(): PermissionStatus {
        return when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU -> {
                // Android 13+ - App-specific directory access
                checkAppSpecificStoragePermission()
            }
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.R -> {
                // Android 11-12 - MANAGE_EXTERNAL_STORAGE
                checkManageExternalStoragePermission()
            }
            else -> {
                // Android 10 and below - WRITE_EXTERNAL_STORAGE
                checkLegacyStoragePermission()
            }
        }
    }
    
    /**
     * Check app-specific storage permission for Android 13+
     */
    private fun checkAppSpecificStoragePermission(): PermissionStatus {
        // For Android 13+, we primarily use app-specific directories
        // which don't require special permissions
        val hasAccess = try {
            val obbDir = context.getExternalFilesDir("obb")
            obbDir?.exists() == true || obbDir?.mkdirs() == true
        } catch (e: Exception) {
            Log.e(TAG, "Error checking app-specific storage access", e)
            false
        }
        
        return PermissionStatus(
            isGranted = hasAccess,
            permissionName = "APP_SPECIFIC_STORAGE",
            rationale = "Required for accessing game OBB files in app-specific directories"
        )
    }
    
    /**
     * Check MANAGE_EXTERNAL_STORAGE permission for Android 11-12
     */
    @RequiresApi(Build.VERSION_CODES.R)
    private fun checkManageExternalStoragePermission(): PermissionStatus {
        val isGranted =
            Environment.isExternalStorageManager()

        return PermissionStatus(
            isGranted = isGranted,
            permissionName = Manifest.permission.MANAGE_EXTERNAL_STORAGE,
            rationale = "Required for full access to OBB files in Android/obb directory",
            requiresSystemSettings = true
        )
    }
    
    /**
     * Check legacy storage permission for Android 10 and below
     */
    private fun checkLegacyStoragePermission(): PermissionStatus {
        val isGranted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
        ) == PackageManager.PERMISSION_GRANTED
        
        return PermissionStatus(
            isGranted = isGranted,
            permissionName = Manifest.permission.WRITE_EXTERNAL_STORAGE,
            rationale = "Required for accessing OBB files in external storage"
        )
    }
    
    /**
     * Request appropriate storage permission based on Android version
     * 
     * @param activity Activity context for permission request
     * @param requestCode Request code for result handling
     */
    fun requestStoragePermission(activity: Activity, requestCode: Int = REQUEST_STORAGE_PERMISSION) {
        when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU -> {
                // Android 13+ - Usually no permission needed for app-specific directories
                Log.d(TAG, "Android 13+ detected - using app-specific storage")
            }
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.R -> {
                // Android 11-12 - Request MANAGE_EXTERNAL_STORAGE
                requestManageExternalStoragePermission(activity, requestCode)
            }
            else -> {
                // Android 10 and below - Request WRITE_EXTERNAL_STORAGE
                requestLegacyStoragePermission(activity, requestCode)
            }
        }
    }
    
    /**
     * Request MANAGE_EXTERNAL_STORAGE permission for Android 11+
     */
    private fun requestManageExternalStoragePermission(activity: Activity, requestCode: Int) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            try {
                val intent = Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION).apply {
                    data = "package:${context.packageName}".toUri()
                }
                activity.startActivityForResult(intent, requestCode)
                Log.d(TAG, "MANAGE_EXTERNAL_STORAGE permission request launched")
            } catch (e: Exception) {
                Log.e(TAG, "Error requesting MANAGE_EXTERNAL_STORAGE permission", e)
                // Fallback to general storage settings
                val fallbackIntent = Intent(Settings.ACTION_MANAGE_ALL_FILES_ACCESS_PERMISSION)
                activity.startActivityForResult(fallbackIntent, requestCode)
            }
        }
    }
    
    /**
     * Request legacy storage permission for Android 10 and below
     */
    private fun requestLegacyStoragePermission(activity: Activity, requestCode: Int) {
        ActivityCompat.requestPermissions(
            activity,
            arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
            requestCode
        )
        Log.d(TAG, "WRITE_EXTERNAL_STORAGE permission request launched")
    }
    
    // ================================================================================================
    // OVERLAY PERMISSION MANAGEMENT
    // ================================================================================================
    
    /**
     * Check overlay permission status
     * 
     * @return PermissionStatus for overlay permission
     */
    fun checkOverlayPermission(): PermissionStatus {
        val isGranted = Settings.canDrawOverlays(context)
        
        return PermissionStatus(
            isGranted = isGranted,
            permissionName = "SYSTEM_ALERT_WINDOW",
            rationale = "Required for floating overlay and mod interface",
            requiresSystemSettings = true
        )
    }
    
    /**
     * Create intent to request overlay permission
     * 
     * @return Intent for requesting overlay permission
     */
    fun createOverlayPermissionIntent(): Intent {
        return Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION).apply {
            data = "package:${context.packageName}".toUri()
        }
    }
    
    /**
     * Request overlay permission
     *
     * @param activity Activity context for permission request
     * @param requestCode Request code for result handling
     */
    fun requestOverlayPermission(activity: Activity, requestCode: Int = REQUEST_OVERLAY_PERMISSION) {
        try {
            val intent = createOverlayPermissionIntent()
            activity.startActivityForResult(intent, requestCode)
            Log.d(TAG, "Overlay permission request launched")
        } catch (e: Exception) {
            Log.e(TAG, "Error requesting overlay permission", e)
        }
    }

    // ================================================================================================
    // BATCH PERMISSION MANAGEMENT
    // ================================================================================================

    /**
     * Check multiple permissions at once
     *
     * @param permissionGroup Group of permissions to check
     * @return Map of permission names to their status
     */
    fun checkBatchPermissions(permissionGroup: String): Map<String, PermissionStatus> {
        val permissions = getPermissionsForGroup(permissionGroup)
        val results = mutableMapOf<String, PermissionStatus>()

        permissions.forEach { permission ->
            results[permission] = when (permission) {
                "UNKNOWN_SOURCES" -> checkUnknownSourcesPermission()
                "STORAGE" -> checkStoragePermission()
                "OVERLAY" -> checkOverlayPermission()
                else -> PermissionStatus(false, permission, "Unknown permission")
            }
        }

        return results
    }

    /**
     * Request multiple permissions in sequence
     *
     * @param activity Activity context for permission requests
     * @param permissionGroup Group of permissions to request
     * @param callback Callback for batch permission results
     */
    fun requestBatchPermissions(
        activity: Activity,
        permissionGroup: String,
        callback: BatchPermissionCallback? = null
    ) {
        val permissions = getPermissionsForGroup(permissionGroup)
        val results = mutableListOf<String>()
        val denied = mutableListOf<String>()
        val permanentlyDenied = mutableListOf<String>()

        Log.d(TAG, "Requesting batch permissions for group: $permissionGroup")

        // Check current status first
        permissions.forEach { permission ->
            val status = when (permission) {
                "UNKNOWN_SOURCES" -> checkUnknownSourcesPermission()
                "STORAGE" -> checkStoragePermission()
                "OVERLAY" -> checkOverlayPermission()
                else -> PermissionStatus(false, permission, "Unknown permission")
            }

            if (status.isGranted) {
                results.add(permission)
            } else {
                denied.add(permission)
            }
        }

        // If all permissions are already granted, return immediately
        if (denied.isEmpty()) {
            callback?.onBatchPermissionResult(
                BatchPermissionResult(
                    allGranted = true,
                    grantedPermissions = results,
                    deniedPermissions = emptyList(),
                    permanentlyDeniedPermissions = emptyList()
                )
            )
            return
        }

        // Request the first denied permission
        // Note: In a real implementation, you might want to implement a queue system
        // for sequential permission requests
        denied.firstOrNull()?.let { permission ->
            when (permission) {
                "UNKNOWN_SOURCES" -> requestUnknownSourcesPermission(activity)
                "STORAGE" -> requestStoragePermission(activity)
                "OVERLAY" -> requestOverlayPermission(activity)
            }
        }
    }

    /**
     * Get list of permissions for a given group
     */
    private fun getPermissionsForGroup(group: String): List<String> {
        return when (group) {
            PERMISSION_GROUP_BASIC -> listOf("OVERLAY")
            PERMISSION_GROUP_STORAGE -> listOf("STORAGE")
            PERMISSION_GROUP_INSTALLATION -> listOf("UNKNOWN_SOURCES")
            PERMISSION_GROUP_ALL -> listOf("UNKNOWN_SOURCES", "STORAGE", "OVERLAY")
            else -> emptyList()
        }
    }

    // ================================================================================================
    // UTILITY METHODS
    // ================================================================================================

    /**
     * Check if all required permissions for BearMod are granted
     *
     * @return true if all essential permissions are granted
     */
    fun areAllRequiredPermissionsGranted(): Boolean {
        val unknownSources = checkUnknownSourcesPermission()
        val storage = checkStoragePermission()
        val overlay = checkOverlayPermission()

        return unknownSources.isGranted && storage.isGranted && overlay.isGranted
    }

    /**
     * Get a comprehensive permission status report
     *
     * @return Formatted string with all permission statuses
     */
    fun getPermissionStatusReport(): String {
        val unknownSources = checkUnknownSourcesPermission()
        val storage = checkStoragePermission()
        val overlay = checkOverlayPermission()

        return buildString {
            appendLine("=== BearMod Permission Status ===")
            appendLine("Unknown Sources: ${if (unknownSources.isGranted) "✓ Granted" else "✗ Denied"}")
            appendLine("Storage Access: ${if (storage.isGranted) "✓ Granted" else "✗ Denied"}")
            appendLine("Overlay Permission: ${if (overlay.isGranted) "✓ Granted" else "✗ Denied"}")
            appendLine("Android Version: ${Build.VERSION.SDK_INT} (API ${Build.VERSION.RELEASE})")
            appendLine("All Required: ${if (areAllRequiredPermissionsGranted()) "✓ Ready" else "✗ Missing"}")
        }
    }

    /**
     * Get user-friendly permission explanation
     *
     * @param permission Permission name to explain
     * @return Human-readable explanation of why the permission is needed
     */
    fun getPermissionExplanation(permission: String): String {
        return when (permission) {
            "UNKNOWN_SOURCES" -> "BearMod needs permission to install game packages (APK files) for proper mod functionality. This allows the app to help you install supported game versions."

            "STORAGE" -> "Storage access is required to read and validate game OBB files. These large data files contain game assets and are essential for proper game operation."

            "OVERLAY" -> "Overlay permission allows BearMod to display its floating interface over games. This is essential for accessing mod features while playing."

            else -> "This permission is required for BearMod to function properly."
        }
    }

    /**
     * Create a user-friendly dialog message for permission requests
     *
     * @param permissions List of permissions to explain
     * @return Formatted message explaining all permissions
     */
    fun createPermissionDialogMessage(permissions: List<String>): String {
        return buildString {
            appendLine("BearMod requires the following permissions to function properly:\n")

            permissions.forEach { permission ->
                val status = when (permission) {
                    "UNKNOWN_SOURCES" -> checkUnknownSourcesPermission()
                    "STORAGE" -> checkStoragePermission()
                    "OVERLAY" -> checkOverlayPermission()
                    else -> PermissionStatus(false, permission, "Unknown")
                }

                val statusIcon = if (status.isGranted) "✓" else "○"
                appendLine("$statusIcon ${status.rationale}")
            }

            appendLine("\nWould you like to grant these permissions now?")
        }
    }

    /**
     * Handle permission result from onActivityResult
     *
     * @param requestCode Request code from the permission request
     * @param resultCode Result code from the activity
     * @param callback Optional callback for permission result
     */
    fun handlePermissionResult(
        requestCode: Int,
        resultCode: Int,
        callback: PermissionCallback? = null
    ) {
        when (requestCode) {
            REQUEST_UNKNOWN_SOURCES -> {
                val status = checkUnknownSourcesPermission()
                if (status.isGranted) {
                    Log.d(TAG, "Unknown Sources permission granted")
                    callback?.onPermissionGranted("UNKNOWN_SOURCES")
                } else {
                    Log.w(TAG, "Unknown Sources permission denied")
                    callback?.onPermissionDenied("UNKNOWN_SOURCES", false)
                }
            }

            REQUEST_STORAGE_PERMISSION, REQUEST_MANAGE_EXTERNAL_STORAGE -> {
                val status = checkStoragePermission()
                if (status.isGranted) {
                    Log.d(TAG, "Storage permission granted")
                    callback?.onPermissionGranted("STORAGE")
                } else {
                    Log.w(TAG, "Storage permission denied")
                    callback?.onPermissionDenied("STORAGE", false)
                }
            }

            REQUEST_OVERLAY_PERMISSION -> {
                val status = checkOverlayPermission()
                if (status.isGranted) {
                    Log.d(TAG, "Overlay permission granted")
                    callback?.onPermissionGranted("OVERLAY")
                } else {
                    Log.w(TAG, "Overlay permission denied")
                    callback?.onPermissionDenied("OVERLAY", false)
                }
            }
        }
    }
}
